server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java

```diff
@@ -644,15 +652,15 @@ public void onFailure(Exception e) {
                         // create failures for all relevant requests
                         for (BulkItemRequest request : requests) {
                             final String indexName = concreteIndices.getConcreteIndex(request.index()).getName();
-                            DocWriteRequest<?> docWriteRequest = request.request();
-                            responses.set(
+                            final DocWriteRequest<?> docWriteRequest = request.request();
+                            final BulkItemResponse bulkItemResponse = new BulkItemResponse(
                                 request.id(),
-                                new BulkItemResponse(
-                                    request.id(),
-                                    docWriteRequest.opType(),
-                                    new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e)
-                                )
+                                docWriteRequest.opType(),
+                                new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e)
                             );
+
+                            indicesService.incrementDocStatusCounter(bulkItemResponse.status());
```

Comment:
Can we do this once at the end of bulk only `incrementDocStatusCounter(int status, int count)`? We can maintain a local map, and only update the concurrent map towards the end of bulk to reduce the concurrent updates to the global map from doc level to request level?

Reply:
Sure, I'll change the path. It will decrease the number of blocking updates to doc counter.